components/ringcentral/sources/common/http-based.js

```diff
@@ -0,0 +1,137 @@
+const template = require("lodash/template");
+const base = require("./base");
+const notificationTypes = require("./notification-types");
+
+module.exports = {
+  ...base,
+  dedupe: "unique",
+  props: {
+    ...base.props,
+    http: {
+      type: "$.interface.http",
+      customResponse: true,
+    },
+  },
+  hooks: {
+    ...base.hooks,
+    async activate() {
+      const opts = {
+        address: this.http.endpoint,
+        eventFilters: this._getEventFilters(),
+      };
+      const {
+        id: webhookId,
+        verificationToken,
+      } = await this.ringcentral.createHook(opts);
+
+      this.db.set("webhookId", webhookId);
+      this.db.set("verificationToken", verificationToken);
+    },
+    async deactivate() {
+      const webhookId = this.db.get("webhookId");
+      await this.ringcentral.deleteHook(webhookId);
+
+      this.db.set("verificationToken", null);
+    },
+  },
+  methods: {
+    ...base.methods,
+    _getPropValues() {
+      return Object.entries(this)
+        .filter(([_, value]) => value != null)
+        .reduce((accum, [prop, value]) => ({
+          ...accum,
+          [prop]: value,
+        }), {});
+    },
+    _getEventFilters() {
+      const eventKeys = this.getSupportedNotificationTypes();
+      return notificationTypes
+        .filter(({ key }) => eventKeys.has(key))
+        .map(({ filter }) => template(filter))
+        .map((templateFn) => templateFn(this._getPropValues()));
+    },
+    /**
+     * Provides the set of notification types to which an HTTP-based event
+     * source subscribes. This should be a subset of the `key` properties
+     * available in the `notification-types` module.
+     *
+     * @return {Set}  The set of supported notification type keys
+     */
+    getSupportedNotificationTypes() {
+      throw new Error("getSupportedNotificationTypes is not implemented");
+    },
+    /**
+     * Validate that the incoming HTTP event comes from the expected source, and
+     * reply with a `200` status code and the proper validation token header, as
+     * described here:
+     * https://community.ringcentral.com/questions/1306/validation-token-is-not-returned-when-creating-a-s.html
+     *
+     * In case the event comes from an unrecognized source, reply with a `404`
+     * status code.
+     *
+     * The result of this method indicates whether the incoming event was valid
+     * or not.
+     *
+     * @param {object}  event The HTTP event that triggers this event source
+     * @return {boolean}  The outcome of the validation check (`true` for valid
+     * events, `false` otherwise)
+     */
+    validateEvent(event) {
+      const {
+        "validation-token": validationToken,
+        "verification-token": verificationToken,
+      } = event.headers;
+
+      const expectedVerificationToken = this.db.get("verificationToken") || verificationToken;
```

Comment:
why are we being lenient about the `verificationToken` not being in db?

Reply:
Whenever we create a hook, RingCentral makes an initial `POST` call to verify that the provided endpoint is valid and healthy. Once it verifies that, the hook creation succeeds. This means that this event source is triggered by an HTTP call in parallel to the execution of the `activate()` method (where the hook is created). The problem is that when the event source is handling that initial HTTP call, the DB is not up-to-date and so we always get a null value when we do `this.db.get("verificationToken")`.

Reply:
Oh that sounds a bit brittle! We might not be great about this use case of webhooks making that initial call an excepting us to be activated already... hmm... There is a chance that initial POST could even fail, right? Is that handled correctly?

Reply:
If that initial POST fails, then the hook creation fails as well and the activation would fail. However, my understanding is that the HTTP endpoint is already set up by the platform before running `activate()`, correct? In which cases could this fail, other than due to a bug on the event source itself?